snorkel/contrib/deps/labeling/model/dep_learner.py

```diff
@@ -0,0 +1,192 @@
+import warnings
+from typing import List, Optional, Tuple
+
+import cvxpy as cp
+import numpy as np
+import scipy as sp
+
+
+class DependencyLearner(object):
+    """A model for learning LF dependencies.
+
+    This class learns the dependencies among labeling functions.
+    It is based on the approach in
+    [Learning Dependency Structures for Weak Supervision Models]
+    (https://arxiv.org/pdf/1903.05844.pdf), published in ICML'19.
+    In this approach, we use robust principal component analysis (rPCA)
+    based algorithm to decompose the inverse covariance matrix into a sparse
+    component that encodes the dependency structure and a low rank component
+    due to marginalizing over the latent true label variable.
+
+    Examples
+    --------
+    >>> dep_learner = DependencyLearner()
+    >>> dep_learner = DependencyLearner(cardinality=3)
+
+    Parameters
+    ----------
+    cardinality
+        Number of classes, by default 2
+
+    Attributes
+    ----------
+    cardinality
+        Number of classes, by default 2
+    """
+
+    def __init__(self, cardinality: int = 2) -> None:
+        self.cardinality = cardinality
+
+    def _force_singleton(
+        self, deps: List[Tuple[int, int]], M: int
+    ) -> List[Tuple[int, int]]:
+        """Force singleton separator assumption given list of dependencies.
+
+        This translates to converting chain dependencies to fully connected clusters of dependencies.
+
+        More information on singleton separator assumption in [AAAI'19](https://arxiv.org/pdf/1810.02840.pdf) and [ICML'19](https://arxiv.org/pdf/1903.05844.pdf) papers.
+        """
+        # remove duplicated pairs
+        deps_singleton = []
+        for i, j in deps:
+            if i < j:
+                deps_singleton.append((i, j))
+
+        # add edges to convert chain deps to cluster deps
+        for i, j in deps:
+            for k, l in deps:
+                if (i == k) and (j < l):
+                    deps_singleton.append((j, l))
+                if (j == l) and (i < k):
+                    deps_singleton.append((i, k))
+                if (j == k) and (i < l):
+                    deps_singleton.append((i, l))
+                if (i == l) and (j < k):
+                    deps_singleton.append((j, k))
+        all_deps = list(set(deps_singleton))
+
+        if len(all_deps) == sp.special.comb(M, 2):
+            raise ValueError(
+                "Dependency structure is fully connected. Rerun with higher thresh_mult."
+            )
+        return all_deps
+
+    def _get_deps_from_inverse_sig(
```

Comment:
How about iterating over j in range(J.shape[0]) and i in range(j)? This way, all pairs will already have i < j and you can skip the _force_singleton first step.